gourmet/gtk/recipe.py

```diff
@@ -0,0 +1,787 @@
+        depth = dest_path.get_depth()
+        if 1 == depth:
+            return True
+        if 2 == depth:
+            parent = dest_path.copy()
+            parent.up()
+            iter = self.get_iter(parent)
+            # Only allow drops onto an ingredient group
+            if not isinstance(self.get_value(iter, 4), _IngredientGroup):
+                return False
+
+            assert 'GTK_TREE_MODEL_ROW' == data.get_data_type().name()
+            success, model, path = _Gtk.tree_get_row_drag_data(data)
+            if success:
+                assert model is self
+                iter = model.get_iter(path)
+                # Only allow ingredients to be dropped onto an ingredient group
+                return not isinstance(self.get_value(iter, 4), _IngredientGroup)
+        return False
+
+
+class _TreeViewSelectionFixer:
+    # Let's try to stop kittens from crying and fix some of GTK's inexplicable
+    # tree view selection behaviors
+
+    def __init__(self, tree, groups):
+        assert isinstance(tree, _Gtk.TreeView)
+
+        # Give keyboard focus to newly-selected rows
+        handler = tree.get_selection().connect('changed', self._focus_on_selected_row)
+
+        # Select the dropped row after drag-and-drop
+        tree.connect('drag-begin', self._start_drag)
+        tree.connect('drag-end', self._end_drag)
+
+        self._tree = tree
+        self._groups = groups
+        self._drag_destination = None
+        self._selection_handler = handler
+        self._insertion_handler = None
+
+
+    # For some reason, GTK doesn't give a newly-selected row the keyboard
+    # focus, so even though a user may click on a row with their mouse, the
+    # keyboard will continue to operate on whatever widget the user was
+    # previously interacting with. So let's fix this.
+    def _focus_on_selected_row(self, selection):
+        model, selected = selection.get_selected()
+        if selected:
+            path = model.get_path(selected)
+            tree = selection.get_tree_view()
+            tree.set_cursor(path, None, None)
+            tree.grab_focus()
+
+
+    # GTK has decided that, upon conclusion of a drag-and-drop operation, the
+    # current row should become whatever row is occupying the location you just
+    # dragged from instead of staying on the dropped row, which is visually
+    # quite confusing: it makes it difficult to determine where the row was
+    # dropped and even whether or not the operation succeeded. So we have to
+    # take care of this ourselves. Of course, GTK doesn't make it easy.
+    # Drag-and-drop is essentially implemented by deleting the row at the drag
+    # point and inserting a new row at the drop point. You might think you
+    # could just connect to the 'row-inserted' signal to select the new row
+    # when it's inserted, but GTK does everything in the opposite order you
+    # might expect: it first inserts a blank row at the drop point -- which is
```

Comment:
```suggestion
    # point and inserting a new row at the drop point. GTK first inserts a 
    # blank row at the drop point -- which is
```
I think that comments should remain neutral and factual.